lib/Perl/Critic/Policy/Subroutines/ProhibitAmbiguousFunctionCalls.pm

```diff
@@ -0,0 +1,169 @@
+package Perl::Critic::Policy::Subroutines::ProhibitAmbiguousFunctionCalls;
+
+use strict;
+use warnings;
+use Readonly;
+use Digest::MD5;
+
+use Perl::Critic::Utils qw{ :severities };
+use base 'Perl::Critic::Policy';
+
+our $VERSION = '1.002';
+
+Readonly::Scalar my $DESC => q{Fully qualified functions calls should end in parens.};
+Readonly::Scalar my $EXPL => q{To differentiate from class methods and function calls, use Foo:Bar::baz()->gimble};
+
+use constant default_severity => $SEVERITY_HIGH;
+use constant default_themes   => qw( core );
+use constant applies_to       => ('PPI::Statement');
+
+use constant supported_parameters => ({
+        name           => 'methods_always_ok',
+        default_string => 'new add',
+        description    => 'Names of methods which never trigger this policy',
+        behavior       => 'string list',
+    },
+    {
+        name           => 'uppercase_module_always_ok',
+        description    => 'Indicates if Foo::Bar->baz is always ok because of the capital F',
+        default_string => 1,
+        behavior       => 'boolean',
+    },
+
+);
+
+my ($docinfo, $file);
+
+sub violates {
+
+    my ($self, $elem, $doc) = @_;
+
+    ## Workaround a slight bug in Perl::Critic
+    return if ref $elem eq 'PPI::Statement::Null';
+
+    ## We may not have a filename, so we need some unique identifier
+    $file = $doc->filename() // Digest::MD5::md5_hex($doc->content());
+
+    ## We need to walk through the whole document, but only the first time we are called per file
+    if (!defined $docinfo->{$file}) {
+        $docinfo->{$file} = {};
+        my $realdoc = $doc->ppi_document;
+        ## This will set elements to matched
+        $self->_kidwalk($realdoc);
+    }
+
+    my $elemid = $self->_nodeid($elem);
+    if (!exists $docinfo->{$file}{$elemid}) {
+        warn "No element found for $elemid";    ## Should not happen
+        return;
+    }
+
+    ## If we have already marked this one, return a violation
+    if ($docinfo->{$file}{$elemid}) {
+        return $self->violation($DESC, $EXPL, $elem);
+    }
+
+    return;
+}
+
+sub _match {
+
+    ## Given a PPI::Node, see if it matches our criteria
+    ## Returns 0 or 1
+
+    my ($self, $lkid) = @_;
+
+    ## We only care about things like Foo::Bar::Baz->gimble
+    return 0 unless $lkid =~ / (\w+(?: ::\w+)+) -> (.+)/x;
+
+    my ($module, $name) = ($1, $2);
```

Comment:
You can combine the two statements above as

```perl
return 0 unless !!( my($module, $name) = $lkid =~ /\w+(?: ::\w+)+) -> (.+)/x );
```

Reply:
True - I'd suggest it's fine as-is, but if you did want a combined version I'd suggest maybe putting the assignment first, since to me that feels like the more important bit:

```
my ($module, $name) = $lkid =~ /\w+(?: ::\w+)+)->(.+)/x
 or return 0;
```

Reply:
Thanks, I'm going to leave as is unless someone has really strong feelings about this. :)